test/functional/p2p_headers_sync_with_minchainwork.py

```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env python3
+# Copyright (c) 2019-2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test that we reject low difficulty headers to prevent our block tree from filling up with useless bloat"""
+
+from test_framework.test_framework import BitcoinTestFramework
+
+import time
+
+NODE1_BLOCKS_REQUIRED = 16
+NODE2_BLOCKS_REQUIRED = 2047
+
+
+class RejectLowDifficultyHeadersTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 3
+        # Node0 has no required chainwork; node1 requires 16 blocks; node2 requires 2048 blocks
+        self.extra_args = [["-minimumchainwork=0x0"], ["-minimumchainwork=0x1f"], ["-minimumchainwork=0x1000"]]
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def test_chains_sync_when_long_enough(self):
+        self.log.info("Generate blocks on the node with no required chainwork")
+        self.generate(self.nodes[0], NODE1_BLOCKS_REQUIRED-2, sync_fun=self.no_op)
+
+        self.log.info("Verify nodes 1 and 2 have no new headers in their headers tree")
+        time.sleep(2)
```

Comment:
I think you could get rid of the `time.sleep` with `assert_debug_log`.
```suggestion
    with self.nodes[1].assert_debug_log(expected_msgs=["[net] Ignoring low-work chain"]), self.nodes[2].assert_debug_log(expected_msgs=["[net] Ignoring low-work chain"]):
         self.generate(self.nodes[0], NODE1_BLOCKS_REQUIRED-2, sync_fun=self.no_op)
```

Reply:
Hm, will `assert_debug_log` work properly here if there might be multiple "[net] Ignoring low-work chain" lines in our debug.log?  (Perhaps that would work if I generated one block at a time in a loop, rather than invoke `generate` once with multiple blocks?)

Reply:
Marking this as closed, I believe this is addressed now with the improved logging and the use of `assert_debug_log`.